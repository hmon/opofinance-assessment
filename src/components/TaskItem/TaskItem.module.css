.TaskItem {
  display: flex;
  position: relative;
  border-bottom: 1px solid rgba(238, 238, 238, 0.2);
  cursor: pointer;
  padding: 0 10px;

  .TaskItemWrapper {
    display: flex;
    flex-direction: row;
    align-items: center;
    width: 100%;
    position: relative;

    @media (max-width: 768px) {
      flex-direction: column;
      align-items: flex-start;
      flex-wrap: wrap;
    }
  }

  .TaskItemContent {
    display: flex;
    flex-direction: column;
    flex-wrap: wrap;
    word-break: break-all;
    width: 100%;
    margin-right: 30px;
    padding: 8px 0 8px 8px;

    .TaskTitle {
      width: max-content;
      font-size: 16px;
      font-weight: 500;
      margin: 0;
      padding: 0;
    }

    .TaskDescription {
      width: max-content;
      color: rgba(255, 255, 255, 0.6);
      font-size: 14px;
      font-weight: 300;
      margin: 0;
      padding: 0;

      @media (max-width: 768px) {
        width: auto;
      }
    }

    .TaskItemDate {
      width: max-content;
      display: flex;
      align-items: center;
      gap: 8px;
      color: rgba(255, 255, 255, 0.6);
      font-size: 14px;
      font-weight: 300;
      margin: 0;
      padding: 0;
    }
  }

  .TaskItemDragHandle {
    position: absolute;
    left: 7px;
    bottom: 7px;
    color: white;
    cursor: move;
  }

  .TaskItemCheckbox {
    --bounding-box-size: 24px;
    all: unset;
    box-sizing: initial;
    cursor: pointer;
    display: grid;
    flex-shrink: 0;
    grid-template-areas: "center" var(--bounding-box-size);
    height: var(--bounding-box-size);
    place-items: center;
    user-select: none;
    width: var(--bounding-box-size);
    margin: 8px 6px 0 -3px;

    .TaskItemCheckboxCircle {
      --circle-diameter: 18px;
      grid-area: center;
      overflow: hidden;
      z-index: 1;
      border: 1px solid rgba(255, 255, 255, 0.6);
      border-radius: 50%;
      width: var(--circle-diameter);
      height: var(--circle-diameter);
    }

    .TaskItemCheckboxCheck {
      color: rgba(255, 255, 255, 0.6);
      grid-area: center;
      opacity: 0;
      pointer-events: none;
      transition: opacity 150ms ease-in-out;
      z-index: 3;
    }
  }

  .TaskItemActions {
    display: flex;
    flex-direction: row;
    justify-content: center;
    align-items: center;
    width: auto;
    gap: 8px;
    padding: 8px 0 8px 8px;

    .TaskItemStatus {
      width: max-content;
      color: rgba(255, 255, 255, 0.6);
      font-size: 14px;
      font-weight: 300;
      margin: 0;
      padding: 0;
      word-break: keep-all;
    }

    .TaskItemDelete {
      all: unset;
      width: max-content;
      padding: 8px;
      cursor: pointer;
      color: rgba(255, 255, 255, 0.6);
      font-size: 14px;
      font-weight: 300;
      border: 1px solid rgba(255, 255, 255, 0.6);
      border-radius: 8px;
      transition: background-color 150ms ease-in-out;

      &:hover {
        background-color: rgba(255, 255, 255, 0.1);
      }
    }

    @media (max-width: 768px) {
      width: 100%;
      justify-content: space-between;
    }
  }

  .TaskTitle, .TaskDescription, .TaskItemDate {
    position: relative;
    z-index: 3;
  }

  .TaskTitle:before, .TaskDescription:before, .TaskItemDate:before {
    content: "";
    display: block;
    position: absolute;
    height: 1px;
    width: 0;
    background-color: white;
    top: calc(50% - 1px);
    left: 0;
    right: 0;
    z-index: 2;
    transition: width 150ms ease-in-out;
  }
}

.TaskItem:hover .TaskItemCheckboxCheck {
  opacity: 1;
}

.TaskItemOverdue {
  background-color: rgba(255, 0, 0, 0.2);
}

.TaskItemCompleted {
  .TaskTitle:before, .TaskDescription:before, .TaskItemDate:before {
    width: 100%;
  }

  .TaskItemCheckboxCheck {
    opacity: 1 !important;
  }
}
