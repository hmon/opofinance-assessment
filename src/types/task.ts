export enum TaskStatus {
  Pending = 'Pending',
  InProgress = 'In Progress',
  Completed = 'Completed',
}

export interface Task {
  id: string;
  title: string;
  description: string;
  status: TaskStatus;
  dueDate: string;
}
