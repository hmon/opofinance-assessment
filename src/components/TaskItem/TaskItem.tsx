import React, { useMemo } from "react";
import { Calendar, Trash } from "react-feather";
import { SortableHandle } from "react-sortable-hoc";
import useTaskStore from "../../store/task.ts";
import { Task, TaskStatus } from "../../types";
import { cx } from "class-variance-authority";
import styles from "./TaskItem.module.css";

interface TaskItemProps {
  task: Task;
}

const DragHandle = SortableHandle(() => (
  <span className={styles.TaskItemDragHandle}>
    <svg width="24" height="24">
      <path fill="currentColor"
            d="M14.5 15.5a1.5 1.5 0 1 1-.001 3.001A1.5 1.5 0 0 1 14.5 15.5zm-5 0a1.5 1.5 0 1 1-.001 3.001A1.5 1.5 0 0 1 9.5 15.5zm5-5a1.5 1.5 0 1 1-.001 3.001A1.5 1.5 0 0 1 14.5 10.5zm-5 0a1.5 1.5 0 1 1-.001 3.001A1.5 1.5 0 0 1 9.5 10.5zm5-5a1.5 1.5 0 1 1-.001 3.001A1.5 1.5 0 0 1 14.5 5.5zm-5 0a1.5 1.5 0 1 1-.001 3.001A1.5 1.5 0 0 1 9.5 5.5z">
      </path>
    </svg>
  </span>
));

const TaskItem: React.FC<TaskItemProps> = React.memo(({
  task
}) => {
  const deleteTask = useTaskStore((state) => state.deleteTask);
  const editTask = useTaskStore((state) => state.editTask);
  const handleDelete = () => deleteTask(task.id);
  const changeStatus = (status: TaskStatus) => editTask({ ...task, status });
  const handleToggleStatus = () => {
    switch (task.status) {
      case TaskStatus.Pending:
        return changeStatus(TaskStatus.InProgress);
      case TaskStatus.InProgress:
        return changeStatus(TaskStatus.Completed);
      case TaskStatus.Completed:
        return changeStatus(TaskStatus.InProgress);
    }
  };

  const classes = cx(styles.TaskItem, {
    [styles.TaskItemOverdue]: new Date(task.dueDate) < new Date(),
    [styles.TaskItemCompleted]: task.status === "Completed"
  });

  const formattedDueDate = useMemo(() => {
    const dueDate = new Date(task.dueDate);

    const intl = new Intl.DateTimeFormat("en-US", {
      year: "numeric",
      month: "long",
      day: "numeric"
    });

    return intl.format(dueDate);
  }, [task.dueDate]);

  return (
    <li
      className={classes}
      onClick={handleToggleStatus}
      data-testid="task-item"
    >
      <DragHandle />

      <button className={styles.TaskItemCheckbox}>
        <span className={styles.TaskItemCheckboxCircle}></span>

        <svg width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg"
             className={styles.TaskItemCheckboxCheck}>
          <path fillRule="evenodd" clipRule="evenodd"
                d="M16.5056 9.00958C16.2128 8.71668 15.7379 8.71668 15.445 9.00958L10.6715 13.7831L8.72649 11.8381C8.43359 11.5452 7.95872 11.5452 7.66583 11.8381C7.37294 12.1309 7.37293 12.6058 7.66583 12.8987L10.1407 15.3736C10.297 15.5299 10.5051 15.6028 10.7097 15.5923C10.8889 15.5833 11.0655 15.5104 11.2023 15.3735L16.5056 10.0702C16.7985 9.77735 16.7985 9.30247 16.5056 9.00958Z"
                fill="currentColor"></path>
        </svg>
      </button>

      <div className={styles.TaskItemWrapper}>
        <div className={styles.TaskItemContent}>
          <div className={styles.TaskTitle}>{task.title}</div>
          <div className={styles.TaskDescription}>{task.description}</div>
          <span className={styles.TaskItemDate}>
          <Calendar size={14}/>
            {formattedDueDate}
        </span>
        </div>

        <div className={styles.TaskItemActions}>
          <div className={styles.TaskItemStatus}>{task.status}</div>
          <button className={styles.TaskItemDelete} data-testid="delete-button" onClick={handleDelete}>
            <Trash size={16}/>
          </button>
        </div>
      </div>
    </li>
  );
});

export default TaskItem;
