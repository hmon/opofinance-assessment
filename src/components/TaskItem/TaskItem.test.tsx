import React from 'react';
import { render, screen } from '@testing-library/react';
import TaskItem from './TaskItem';
import useTaskStore from "../../store/task";
import { TaskStatus } from "../../types";

const mockTask = {
  id: '1',
  title: 'Foo Task',
  description: 'Foo Description',
  status: TaskStatus.InProgress,
  dueDate: new Date('2022-01-01').toDateString(),
};

describe('TaskItem', () => {
  test('renders TaskItem component', () => {
    render(<TaskItem task={mockTask} />);
    const taskItemElement = screen.getByTestId('task-item');
    expect(taskItemElement).toBeInTheDocument();
    expect(taskItemElement).toMatchSnapshot();
  });

  test('calls handleDelete when delete button is clicked', () => {
    useTaskStore.getState().deleteTask = vi.fn();
    render(<TaskItem task={mockTask} />);
    const deleteButton = screen.getByTestId('delete-button');
    expect(deleteButton).toBeInTheDocument();
    deleteButton.click();
    expect(useTaskStore.getState().deleteTask).toHaveBeenCalledWith('1');
  });
});
