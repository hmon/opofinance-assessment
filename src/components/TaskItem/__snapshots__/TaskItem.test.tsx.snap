// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TaskItem > renders TaskItem component 1`] = `
<li
  class="_TaskItem_a9f61b _TaskItemOverdue_a9f61b"
  data-testid="task-item"
>
  <span
    class="_TaskItemDragHandle_a9f61b"
  >
    <svg
      height="24"
      width="24"
    >
      <path
        d="M14.5 15.5a1.5 1.5 0 1 1-.001 3.001A1.5 1.5 0 0 1 14.5 15.5zm-5 0a1.5 1.5 0 1 1-.001 3.001A1.5 1.5 0 0 1 9.5 15.5zm5-5a1.5 1.5 0 1 1-.001 3.001A1.5 1.5 0 0 1 14.5 10.5zm-5 0a1.5 1.5 0 1 1-.001 3.001A1.5 1.5 0 0 1 9.5 10.5zm5-5a1.5 1.5 0 1 1-.001 3.001A1.5 1.5 0 0 1 14.5 5.5zm-5 0a1.5 1.5 0 1 1-.001 3.001A1.5 1.5 0 0 1 9.5 5.5z"
        fill="currentColor"
      />
    </svg>
  </span>
  <button
    class="_TaskItemCheckbox_a9f61b"
  >
    <span
      class="_TaskItemCheckboxCircle_a9f61b"
    />
    <svg
      class="_TaskItemCheckboxCheck_a9f61b"
      fill="none"
      height="24"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M16.5056 9.00958C16.2128 8.71668 15.7379 8.71668 15.445 9.00958L10.6715 13.7831L8.72649 11.8381C8.43359 11.5452 7.95872 11.5452 7.66583 11.8381C7.37294 12.1309 7.37293 12.6058 7.66583 12.8987L10.1407 15.3736C10.297 15.5299 10.5051 15.6028 10.7097 15.5923C10.8889 15.5833 11.0655 15.5104 11.2023 15.3735L16.5056 10.0702C16.7985 9.77735 16.7985 9.30247 16.5056 9.00958Z"
        fill="currentColor"
        fill-rule="evenodd"
      />
    </svg>
  </button>
  <div
    class="_TaskItemWrapper_a9f61b"
  >
    <div
      class="_TaskItemContent_a9f61b"
    >
      <div
        class="_TaskTitle_a9f61b"
      >
        Foo Task
      </div>
      <div
        class="_TaskDescription_a9f61b"
      >
        Foo Description
      </div>
      <span
        class="_TaskItemDate_a9f61b"
      >
        <svg
          fill="none"
          height="14"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="14"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            height="18"
            rx="2"
            ry="2"
            width="18"
            x="3"
            y="4"
          />
          <line
            x1="16"
            x2="16"
            y1="2"
            y2="6"
          />
          <line
            x1="8"
            x2="8"
            y1="2"
            y2="6"
          />
          <line
            x1="3"
            x2="21"
            y1="10"
            y2="10"
          />
        </svg>
        January 1, 2022
      </span>
    </div>
    <div
      class="_TaskItemActions_a9f61b"
    >
      <div
        class="_TaskItemStatus_a9f61b"
      >
        In Progress
      </div>
      <button
        class="_TaskItemDelete_a9f61b"
        data-testid="delete-button"
      >
        <svg
          fill="none"
          height="16"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polyline
            points="3 6 5 6 21 6"
          />
          <path
            d="M19 6v14a2 2 0 0 1-2 2H7a2 2 0 0 1-2-2V6m3 0V4a2 2 0 0 1 2-2h4a2 2 0 0 1 2 2v2"
          />
        </svg>
      </button>
    </div>
  </div>
</li>
`;
